### PR TITLE
Fix clippy warnings

### PR DESCRIPTION
**Pull request by TianyiShi2001**
Fixing these clippy warnings makes the code cleaner and more readable, and in some cases improves performance (particularly, fixing unnecessary lazy evaluation by changing `ok_or_else` to `ok_or` (well, maybe not, because the compiler may be smart enough to optimise them out)).

### CODE REVIEW

File: src/io/free_functions.rs [hunk 1]
```diff
@@ -235,8 +235,7 @@ pub(crate) fn guess_format_from_path_impl(path: &Path) -> Result<ImageFormat, Pa
         .and_then(|s| s.to_str())
         .map(str::to_ascii_lowercase);
 
-    let ext = ext.as_ref()
-        .map(String::as_str);
+    let ext = ext.as_deref();
```
Comment (kaj) [inline src/io/free_functions.rs:238]: I think this requires rust 1.40.0, while `image` still promises backwards compatibility to 1.36.0.

File: src/pnm/decoder.rs [hunk 2]
```diff
@@ -616,10 +616,7 @@ impl<R: Read> PnmDecoder<R> {
 fn read_separated_ascii<T: FromStr<Err = ParseIntError>>(reader: &mut dyn Read) -> ImageResult<T>
     where T::Err: Display
 {
-    let is_separator = |v: &u8| match *v {
-        b'\t' | b'\n' | b'\x0b' | b'\x0c' | b'\r' | b' ' => true,
-        _ => false,
-    };
+    let is_separator = |v: &u8| matches! { *v, b'\t' | b'\n' | b'\x0b' | b'\x0c' | b'\r' | b' '};
```
Comment (kaj) [inline src/pnm/decoder.rs:619]: Very nice, but requires rust 1.42.0 (of course, this is about all uses of `matches!`).

**Review by kaj — COMMENTED** [review]
Nice stuff, but some requires rust 1.40 or 1.42, and I think image still promises compatibility with rust 1.36.

File: src/io/free_functions.rs (hunk 1 above)
Comment (HeroicKatora) [reply src/io/free_functions.rs:238]: With `1.34.2` to be exact.

**Comment by TianyiShi2001** [comment]
I've reverted the incompatible fixes (will redo them at some point in the future). Can the remaining fixes be merged?

File: src/io/free_functions.rs (hunk 1 above)
Comment (TianyiShi2001) [reply src/io/free_functions.rs]: Thank you! I've reverted this.

File: src/pnm/decoder.rs (hunk 2 above)
Comment (TianyiShi2001) [reply src/pnm/decoder.rs]: Thank you! I've reverted this.

**Comment by HeroicKatora** [comment]
Can you revert them by rebasing and removing the commits and then force pushing instead? Github is weird in that it closes pending PRs if they only contain commits of another PR.. Regardless of whether that other PR contains commits that revert them. We can also squash merge.

> Can the remaining fixes be merged?

Yes, I think so.

> Will redo them at some point in the future

If you open another PR with only those fixes then I'll tag it appropriately. The number of good changes that would require a new MSRV is a factor in the consideration of when to bump it.

**Review by kaj — APPROVED** [review]

**Comment by TianyiShi2001** [comment]
> Can you revert them by rebasing and removing the commits and then force pushing instead? 

Thank you! This has been done.

**Review by HeroicKatora — APPROVED** [review]

**Comment by HeroicKatora** [comment]
@TianyiShi2001 I see you're becoming quite interested here and you already have a few PRs merged now. Time for a formal invite. Hi :wave: [The collaborator role is explained here](https://github.com/image-rs/organization/blob/master/CONTRIBUTING.md#membership-and-teams). Basically, you can merge/tag/review PRs and you only have to stick to the code of conduct.

**Pull request merged**